core/src/main/java/org/apache/druid/utils/CollectionUtils.java

```diff
@@ -88,12 +89,22 @@ public int size()
 
   /**
    * Returns a transformed map from the given input map where the key is modified based on the given keyMapper
-   * function.
+   * function. This method fails if keys collide after applying the  given keyMapper function and
+   * throws a IllegalStateException.
+   *
+   * @throws ISE if key collisions occur while applying specified keyMapper
    */
+
   public static <K, V, K2> Map<K2, V> mapKeys(Map<K, V> map, Function<K, K2> keyMapper)
   {
     final Map<K2, V> result = Maps.newHashMapWithExpectedSize(map.size());
-    map.forEach((k, v) -> result.put(keyMapper.apply(k), v));
+    map.forEach((k, v) -> {
+      final K2 k2 = keyMapper.apply(k);
+      if (result.containsKey(k2)) {
```

Comment:
Please make just one lookup operation for every key: putIfAbsent().

Reply:
@surekhasaharan the idea of this replacement was to get rid of `containsKey()` call, so that only one lookup operation (`putIfAbsent()` itself) is done on each iteration instead of two (`containsKey()` and `put()`):
```java
if (result.putIfAbsent(k2, v) != null) {
  throw new ISE("Conflicting key[%s] calculated via keyMapper for original key[%s]", k2, k);
}
```

Reply:
i see, fixed.